src/sagemaker/jumpstart/artifacts/model_uris.py

```diff
@@ -84,14 +87,55 @@ def _retrieve_model_uri(
         sagemaker_session=sagemaker_session,
     )
 
+    model_artifact_key: str
+
     if model_scope == JumpStartScriptScope.INFERENCE:
+        instance_specific_prepacked_hosting_artifact_key: Optional[str] = (
+            model_specs.hosting_instance_type_variants.get_instance_specific_prepacked_artifact_key(
+                instance_type=instance_type
+            )
+            if instance_type
+            and getattr(model_specs, "hosting_instance_type_variants", None) is not None
+            else None
+        )
+
+        instance_specific_hosting_artifact_key: Optional[str] = (
+            model_specs.hosting_instance_type_variants.get_instance_specific_artifact_key(
+                instance_type=instance_type
+            )
+            if instance_type
+            and getattr(model_specs, "hosting_instance_type_variants", None) is not None
+            else None
+        )
+
+        default_prepacked_hosting_artifact_key: Optional[str] = getattr(
+            model_specs, "hosting_prepacked_artifact_key"
+        )
+
+        default_hosting_artifact_key: str = model_specs.hosting_artifact_key
+
         model_artifact_key = (
-            getattr(model_specs, "hosting_prepacked_artifact_key", None)
-            or model_specs.hosting_artifact_key
+            instance_specific_prepacked_hosting_artifact_key
+            or instance_specific_hosting_artifact_key
+            or default_prepacked_hosting_artifact_key
+            or default_hosting_artifact_key
         )
```

Comment:
The hosting artifact retrieval logic is quite complex and the implementation is too linear IMO. Could you break it down into utilities that you can unit test separately.

E.g.
```py
if model_scope == JumpStartScriptScope.INFERENCE:
  if is_model_prepacked(...):
    return artifacts.retrieve_instance_specific_hosting_prepack_key(...)
  else:
    return artifacts.retrieve_instance_specific_hosting_artifact_key(...)
elif model_scope === JumpStartScriptScope.TRAINING:
  return artifacts.retrieve_instance_specific_training_artifact_key(...)
```

Then delegate to each utility method the task of handling whether (a) an instance was passed, (b) if (a) is true, whether it maps + add unit tests for each case.